Add troubleshooting page for unsupported operations in Mountpoint (#703)
awslabs/mountpoint-s3: doc/TROUBLESHOOTING.md

> # Troubleshooting

Reviewer: just as a possible source of inspiration for this page, I really liked how EKS structured their doc:
https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html

basically, the problem/solution sections is a good template for all of the issues; not sure though, that all of the problems described in our doc have a solution/workaround, but it would be nice to think in that direction

Author: In our doc, we are only adding logs for unsupported operations. So, they do not have any solutions. It is for users to know that it is expected behaviour.
Although, for some operations we can add solution like `allow-overwrite` and `allow-delete`.

Reviewer 2: While we're starting with unsupported operations, I do like the EKS doc. I think you already have a structure very similar, Ankit, so we're on the right track!

and to write new objects sequentially from a single client at a time.
To achieve this, Mountpoint does not implement all the features of a POSIX file system that may affect compatibility with your application.

This documentation enumerates some examples of what error messages may look like when trying to perform unsupported operations. Please also take a look at Mountpoint's [semantics documentation](../doc/SEMANTICS.md) for more more detailed information on the decisions and tradeoffs made.
This troubleshooting page is based on release Mountpoint-s3 v1.4.0 .Versions earlier to that might not have same error logging.

## Random Write

Mountpoint supports writing to a file sequentially. Random writes, or 'out-of-order' writes, will return an error to FUSE, which may appear in applications with the error message "Invalid argument".
For example, the following code seeks one byte into the file and writes a single byte using `dd`.

```
$ dd if=/dev/random of=out seek=1 count=1
dd: writing to 'out': Invalid argument
```

In Mountpoint's logs, a warning message will be emitted similar to below:

```
WARN write{req=52 ino=49 fh=3 offset=512 length=512 name="out"}: 

> mountpoint_s3::fuse: write failed:upload error: out of order write NOT supported by Mountpoint, aborting the upload; expected offset 0 but got 512

Author: Yes. Earlier I was breaking the line at that point. In modifying that, it got malformed.

```

## Writing to an existing File 

Mountpoint support overwriting to existing file using mount option `--allow-overwrite` . 
Trying to open an existing file for writing without `--allow-overwrite`  flag will fail with the error: `Operation not permitted`.
For example, there is an pre-existing file 'existing-file.txt' in mounted directory. We try to overwrite the 'existing-file' as follow:

```
$ echo "Overwriting a file..." > existing-file.txt
operation not permitted: existing-file.txt
```

Log entries for file overwriting looks like:

```
WARN setattr{req=4 ino=21 name="existing-file.txt"}: 
mountpoint_s3::fuse: setattr failed: inode error: inode 21 (full key "existing-file.txt") is a remote inode and its attributes cannot be modified
```

> ### Unreleased

Reviewer: this is hard to follow, may be put here only the final version and merge this page when the new release was created? probably, leaving a note like the following for prev versions, though:

> Note: in versions of MP before 1.5.0 the error will look like `mountpoint_s3::fuse: setattr failed: inode error: inode 21 (full key "existing-file.txt") is a remote inode and its attributes cannot be modified`

Author: I think we can merge this troubleshooting page for now and make the change once the new release is made. Because we don't know when are we going to do next patch release. If it is soon enough, I think we can wait.

Reviewer 3: It might be helping users who build from source but also confusing for users that use official releases. I would not include it in the doc for now.

Author: In that case, I think we can keep troubleshooting page updated with #718 and once that PR gets released in something like v1.4.1, we can merge this PR (soon after or before??)
Also, how do we make sure that this PR is ready to be merged but not merge it till next release? Just approve it and wait? We might need to rebase (although unlikely) and be careful to not merge it before the release.

Reviewer: Yeah, its frustrating to have PRs open for too long. You could put something like this to unblock the merge:
___

Log entries for file overwrites looks like one of the following depending on the Mountpoint version:
```
WARN setattr{req=4 ino=21 name="existing-file.txt"}: 
mountpoint_s3::fuse: setattr failed: inode error: inode 21 (full key "existing-file.txt") is a remote inode and its attributes cannot be modified
```
```
WARN setattr{req=11 ino=2 name="existing-file.txt"}: 
mountpoint_s3::fuse: setattr failed: file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it
```
___

It's just this `Unreleased` looks a little off here.

With improvement in error logging, we will get the following logs for file overwrite-

In the logs, we get the following WARN message:

```
WARN setattr{req=11 ino=2 name="existing-file.txt"}: 
mountpoint_s3::fuse: setattr failed: file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it
```

> ## Deleting file

Reviewer 2: ```suggestion
## Deleting files
```

> In order to delete files using Mountpoint user need to enable it using `--allow-delete` CLI flag. To know more about the behaviour of file deletion, please visit [Delete Semantics](https://github.com/awslabs/mountpoint-s3/blob/main/doc/SEMANTICS.md#deletes)

Reviewer 2: As much as I hate to replace British English with American English... :)

```suggestion
In order to delete files using Mountpoint, you must opt-in using the `--allow-delete` CLI flag.
To learn more about how file deletion works in Mountpoint, please review the [file deletion section of Mountpoint's semantics documentation](https://github.com/awslabs/mountpoint-s3/blob/main/doc/SEMANTICS.md#deletes).
```

Without the flag, if a customer tries to delete, for example test-file.txt, they will get the following error:

```
$ rm test-file.txt
rm: cannot remove 'test-file.txt': Operation not permitted
```

In Mountpoint's logs, following message will be emitted:

```
WARN unlink{req=8 parent=1 name="test-file.txt"}: 
mountpoint_s3::fuse: unlink failed: Deletes are disabled. Use '--allow-delete' mount option to enable it.
```

## Listing file which is shadowed by directory of same name

A file system directory cannot contain both a file and a directory of the same name. If your bucket's directory structure would result in this state, only the directory will be accessible. The object will not be accessible. 
For more details on how Mountpoint maps S3 object keys to files and directories, see the [semantics documentation](https://github.com/awslabs/mountpoint-s3/blob/main/doc/SEMANTICS.md#mapping-s3-object-keys-to-files-and-directories). 

> For example, our bucket contains the following object keys:

Reviewer 2: new line needed

```suggestion

For example, if a bucket contains the following object keys:
```

* out

> * out/image.jpg

Reviewer 2: Unless it makes sense (like `out` file), let's use things like `a`/`b` rather than `out`.

When listing the content of a directory mounting the bucket in the example above, we will get the following result:

```
$ ls
out
```

Mountpoint logs will show an entry like this:

> ```
> WARN readdir{req=5 ino=1 fh=2 offset=17}: 
> mountpoint_s3::inode::readdir::ordered:file 'out' (full key "out") is omitted because another directory 'out' exist with the same name.
> ```

Reviewer 2: malformed?

````suggestion
```
WARN readdir{req=5 ino=1 fh=2 offset=17}: 
mountpoint_s3::inode::readdir::ordered: file 'out' (full key "out") is omitted because another directory 'out' exist with the same name.
```
````

> Although, in S3 express one zone bucket both file and directory will be shown without the above warning being emitted. 

Reviewer 2: That's a bug? If its a bug, we should open an issue and reference it.

We should also capitalize the name "S3 Express One Zone".

## Renaming a file/directory

Renaming a file or a directory inside the mounted directory is not supported by Mountpoint.
Attempting to rename will give error like: 

```
$ mv hello.txt hello_new.txt
mv: rename hello.txt to hello_new.txt: Function not implemented
```

Mountpoint logs will show following message for the above error:

```
rename{req=2 parent=1 name="" newparent=1 newname=""}: mountpoint_s3::fuse: rename failed: operation not supported by Mountpoint

> ```

Reviewer 2: Is this example right? `name` and `newname` will really be blank?

Author: No, it contains both the names in Linux. I was checking all the error messages from Ubuntu 22.04, but somehow I checked this from MacOS. I should update it.

## Accessing Glacier objects

Objects in the Glacier Flexible Retrieval and Glacier Deep Archive storage classes, and the Archive Access and Deep Archive Access tiers of S3 Intelligent-Tiering, are not accessible with Mountpoint. To access these objects with Mountpoint, restore or copy them to another storage class first.
When listing or trying to access objects in these storage classes, Mountpoint logs will show entries like:

```
WARN readdirplus{req=10 ino=1 fh=1 offset=0}: 
mountpoint_s3::inode: objects in the GLACIER and DEEP_ARCHIVE storage classes are only accessible if restored

> ```

Reviewer 2: Why do we warn on `readdirplus`? Should we fix that?

Author: This error is propagated from `is_readable()` [method](https://github.com/awslabs/mountpoint-s3/blob/92aec0a0132aaf8d20093943cb40269bb57ffb2f/mountpoint-s3/src/inode.rs#L1488-L1503) implemented on InodeStat.

Reviewer 2: Ah I see... I'd probably just include one of these two logs since they are very similar.

For more detailed information, please refer to Mountpoint's [semantics documentation](../doc/SEMANTICS.md).
This document aims to capture errors from the latest versions of Mountpoint.
If you are using an older version of Mountpoint, please refer to older versions of this document.

## Random writes

Mountpoint supports writing to a file sequentially. Random writes, or 'out-of-order' writes, will return an error to FUSE which may appear in applications with the error message "Invalid argument".
For example, the following code seeks one byte into the file and writes a single byte using `dd`.

```
$ dd if=/dev/random of=out seek=1 count=1
dd: writing to 'out': Invalid argument
```

In Mountpoint's logs, a warning message will be emitted similar to below:

```
WARN write{req=52 ino=49 fh=3 offset=512 length=512 name="out"}: 
mountpoint_s3::fuse: write failed: upload error: out of order write NOT supported by Mountpoint, aborting the upload; expected offset 0 but got 512
```

## Writing to an existing file 

Trying to open an existing file for writing using Mountpoint without `--allow-overwrite`  flag will fail with the error: `Operation not permitted`.

> For example: for a directory with existing file 'existing-file.txt', the following error will occur when trying to overwrite it:

Reviewer 3: ```suggestion
For example, overwriting an existing file in Mountpoint would result in the following error:
```

# Troubleshooting

Mountpoint for Amazon S3 is optimized for applications that need high read throughput to large objects, potentially from many clients at once,
and to write new objects sequentially from a single client at a time.
To achieve this, Mountpoint does not implement all the features of a POSIX file system that may affect compatibility with your application.

This documentation enumerates some examples of what error messages may look like when trying to perform unsupported operations. Please also take a look at Mountpoint's [semantics documentation](../doc/SEMANTICS.md) for more more detailed information on the decisions and tradeoffs made.

> This troubleshooting page is based on release Mountpoint-s3 v1.4.0 .Versions earlier to that might not have same error logging.

Reviewer 3: ```suggestion
This troubleshooting page is based on Mountpoint for Amazon S3 v1.4.0 or newer. Versions earlier to that might not have same error logging.
```

```
$ ls mnt/
out
```

Mountpoint logs will show an entry like this:

```
WARN readdir{req=5 ino=1 fh=2 offset=17}: 
mountpoint_s3::inode::readdir::ordered:file 'out' (full key "out") is omitted because another directory 'out' exist with the same name.
```

Although, in S3 express one zone bucket both file and directory will be shown without the above warning being emitted. 

For more details on how Mountpoint maps S3 object keys to files and directories, see the [semantics documentation](https://github.com/awslabs/mountpoint-s3/blob/main/doc/SEMANTICS.md#mapping-s3-object-keys-to-files-and-directories). 

## Renaming a file/directory

Renaming a file or a directory inside the mounted directory is not supported by Mountpoint.
Attempting to rename will give error like: 

```
$ mv hello.txt hello_new.txt
mv: rename hello.txt to hello_new.txt: Function not implemented
```

Mountpoint logs will show following message for the above error:

```
rename{req=2 parent=1 name="" newparent=1 newname=""}: mountpoint_s3::fuse: rename failed: operation not supported by Mountpoint
```

## Accessing Glacier objects

> Objects in the Glacier Flexible Retrieval and Glacier Deep Archive storage classes, and the Archive Access and Deep Archive Access tiers of S3 Intelligent-Tiering, are not accessible with Mountpoint. To access these objects with Mountpoint, restore or copy them to another storage class first.

Reviewer 3: maybe
```suggestion
Objects in Glacier Flexible Retrieval storage class, Glacier Deep Archive storage class, and non-instant access tiers of S3 Intelligent-Tiering storage class are not accessible with Mountpoint. To access these objects with Mountpoint, restore or copy them to another storage class first. See more details in [aws documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/archived-objects.html).
```

Reviewer 2: nice! just capitalize AWS

When listing or trying to access objects in these storage classes, Mountpoint logs will show entries like:

```
WARN readdirplus{req=10 ino=1 fh=1 offset=0}: 
mountpoint_s3::inode: objects in the GLACIER and DEEP_ARCHIVE storage classes are only accessible if restored
```

```
WARN lookup{req=6 ino=1 name="class_GLACIER"}: 
mountpoint_s3::inode: objects in the GLACIER and DEEP_ARCHIVE storage classes are only accessible if restored
```

## Modifying metadata

> Mountpoint does not support modifying metadata such as file modification times and file size. If user try to modify, for example file modification time using `touch`, user will get following error:

Reviewer 3: ```suggestion
Mountpoint does not support modifying metadata such as file modification times and file size. If users try to modify them, for example, updating file modification time with `touch`, user will get following error:
```

 
```
$ touch -a -m -t 201512180130.09 init.txt
touch: init.txt: Operation not permitted
```

> And the Mountpoint Logs will show:

Reviewer 3: probably
```suggestion
And Mountpoint logs will show:
```

# Troubleshooting Production Issues related to Unsupported Operations in Mountpoint

Mountpoint enables users to access S3 buckets through local filesystem. While this allows for existing workflows and applications to work with S3 unchanged, it also limits the effective API for Mountpoint to be a subset of the filesystem API.
So, there are a few operations in filesystem API which are not supported using Mountpoint on a mounted bucket. Here, we are enumerating those common cases with their errors reported in the logs which customers face. Please look at [SEMANTICS document](doc/SEMANTICS.md) for more information on what operations are allowed in Mountpoint.

> ## Non-sequential Write
> Mountpoint only supports sequential writes. On random out of order writes users should expect `Invalid Argument`. For example - 
> ```
> $ dd if=/dev/random of=out seek=1 count=1
> dd: writing to 'out': Invalid argument
> ```
> In the logs, we get the following WARN message - 
> ```
> WARN write{req=52 ino=49 fh=3 offset=512 length=512 name="out"}: mountpoint_s3::fuse: write failed:\
> upload error: out of order write NOT supported by Mountpoint, aborting the upload; expected offset 0 but got 512
> ```

Reviewer 2: Let's format the markdown with more spacing around headers and code blocks.

Also, we should prefer passive voice in the documentation where possible.

````suggestion
## Random writes

Mountpoint supports writing to a file sequentially. Random writes, or 'out-of-order' writes, will return an error to FUSE, which may appear in applications with the error message "Invalid argument".

For example, the following code seeks one byte into the file and writes a single byte using `dd`.

```
$ dd if=/dev/random of=out seek=1 count=1
dd: writing to 'out': Invalid argument
```

In Mountpoint's logs, a warning message will be emitted similar to below:

```
WARN write{req=52 ino=49 fh=3 offset=512 length=512 name="out"}: mountpoint_s3::fuse: write failed: upload error: out of order write NOT supported by Mountpoint, aborting the upload; expected offset 0 but got 512
```
````

## Writing to an existing File 
Mountpoint does support overwriting to existing file using mount option `--allow-overwrite` . Trying to open an existing file for writing without `--allow-overwrite`  flag will fail with the error: `Operation not permitted`.
In the logs, we get the following WARN message - 
```

> WARN file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it

Author: I am not sure what example to give for the case of opening file not in truncate mode. Adding a code snippet to do that will differ from other cases in this document. Let me dive deeper if there is any command line method to do it.

> ```

Reviewer 2: I liked the example you gave above with `dd`. Can you have one with something like...:

```sh
echo "Overwriting a file..." > /mnt/s3-bucket/existing-file.txt
```

nit: "Mountpoint supports" instead of "Mountpoint does support"

Author: echo gives different error -
`WARN setattr{req=12 ino=2 name="file.txt"}: mountpoint_s3::fuse: setattr failed: inode error: inode 2 (full key "file.txt") is a remote inode and its attributes cannot be modified`

Author: I was trying to use vim. But, it is able to open the file using vim. I am just not able to overwrite. Not sure if I should add that as an example.

Author: I have updated the error according to my new PR #718 for the case of overwrite and modifying metadata.

> ## Deleting file (`rm` without mount option `--allow-delete`)

Reviewer 2: ditch the bit in brackets

In order to delete files using Mountpoint user need to enable it using `--allow-delete` mount option. To know more about the behaviour of file deletion, please visit [Delete Semantics](https://github.com/awslabs/mountpoint-s3/blob/main/doc/SEMANTICS.md#deletes)

> Without the flag, if a customer tries to run  `rm test-file.txt` command, they will see the following error: `rm: cannot remove 'test-file.txt': Operation not permitted`

Reviewer 2: Show a full example as a separate block.

In the logs, this will show up as 
```
WARN unlink{req=8 parent=1 name="test-file.txt"}: mountpoint_s3::fuse: unlink failed: Deletes are disabled. \
Use '--allow-delete' mount option to enable it.
```

## Listing file which is shadowed by directory of same name

> A file system directory cannot contain both a file and a directory of the same name. If your bucket's directory structure would result in this state, only the directory will be accessible. For example, our bucket contains the following object keys:

Reviewer 2: Is this true? What happens with S3 Express One Zone?

Author: Yes. It is true for normal bucket. But, for S3 Express One zone, we are able to see both the file and directory. Also, this warning is not emitted. 
But, still reading the file is not possible. It gives error - 
`cat: out: Is a directory`

```

## Moving a file/directory
Mountpoint supports moving files from another filesystem into the mounted directory. But, renaming a file or a directory inside the mounted directory is not supported by Mountpoint and attempting it will give error like `mv: rename hello.txt to hello_new.txt: Function not implemented`.
The logs will show following message for the above error - 
```
rename{req=2 parent=1 name="" newparent=1 newname=""}: mountpoint_s3::\
fuse: rename failed: operation not supported by Mountpoint
```

## Accessing Glacier Objects
Objects in the Glacier Flexible Retrieval and Glacier Deep Archive storage classes, and the Archive Access and Deep Archive Access tiers of S3 Intelligent-Tiering, are not accessible with Mountpoint. To access these objects with Mountpoint, restore or copy them to another storage class first.
When listing or trying to access objects in these storage classes, the logs will show entries like:
```
WARN readdirplus{req=10 ino=1 fh=1 offset=0}: mountpoint_s3::inode: objects in \
the GLACIER and DEEP_ARCHIVE storage classes are only accessible if restored
```

> ## Modifying file modification time and Sizes

Reviewer 3: We don't allow modifying metadata in general, so maybe
```suggestion
## Modifying metadata
```
